api/src/pcapi/core/token/__init__.py

```diff
@@ -0,0 +1,68 @@
+from datetime import datetime
+from datetime import timedelta
+import enum
+
+from flask import current_app as app
+
+from pcapi.core.users import utils
+from pcapi.core.users.exceptions import InvalidToken
+
+
+class TokenType(enum.Enum):
+    EMAIL_CHANGE_CONFIRMATION = "update_email_confirmation"
+    EMAIL_CHANGE_VALIDATION = "update_email_validation"
+
+
+class Token:
+    token_type: TokenType
+    user_id: int
+    encoded_token: str
+    data: dict
+
+    def __init__(self, encoded_token: str):
+        self.encoded_token = encoded_token
+        try:
+            payload = utils.decode_jwt_token(encoded_token)
+            self.data = payload["data"] if payload["data"] is not None else {}
+            self.token_type = TokenType(payload["token_type"])
+            self.user_id = payload["user_id"]
+        except Exception:
+            raise InvalidToken()
+
+    def _get_token_key(self) -> str:
+        return get_token_key(self.token_type, self.user_id)
+
+    def get_expiration_date(self) -> datetime | None:
+        ttl = app.redis_client.ttl(self._get_token_key())  # type: ignore [attr-defined]
+        if ttl < 0:
+            # -2 if doesn't exist, -1 if no expiration
+            return None
+        return datetime.utcnow() + timedelta(seconds=ttl)
+
+    def check_token(self, token_type: TokenType, user_id: int | None = None) -> None:
+        if (
+            not app.redis_client.exists(self._get_token_key())  # type: ignore [attr-defined]
+            or app.redis_client.get(self._get_token_key()) != self.encoded_token  # type: ignore [attr-defined]
+            or self.token_type != token_type
+            or (user_id is not None and self.user_id != user_id)
+        ):
+            raise InvalidToken()
+
+    def expire_token(self) -> None:
+        app.redis_client.delete(self._get_token_key())  # type: ignore [attr-defined]
+        self.expiration_date = None
+
+
+def create_token(token_type: TokenType, ttl: timedelta | None, user_id: int, data: dict | None = None) -> Token:
+    encoded_token = utils.encode_jwt_payload({"token_type": token_type.value, "user_id": user_id, "data": data})
+    token = Token(encoded_token)
+    app.redis_client.set(token._get_token_key(), encoded_token, ex=ttl)  # type: ignore [attr-defined]
+    return token
+
+
+def get_token_key(token_type: TokenType, user_id: int) -> str:
```

Comment:
(détails) Je trouve qu'inclure le mot Redis dans le nom de la fonction la rendrait plus explicite : 

```suggestion
def get_redis_key(token_type: TokenType, user_id: int) -> str:
```

Micro-détail : cette fonction étant à usage interne de ce module, on pourrait la préfixer par `_`.